# Exploring eBay Car Sales Data

In [74]:
import pandas as pd
import numpy as np
from datetime import datetime
# import seaborn as sns
# import matplotlib.pyplot as plt
# %matplotlib inline

autos = pd.read_csv('autos.csv', encoding='Latin-1')

## Exploration

In [75]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [76]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [77]:
nulls = {}
for row in autos.columns:
    nulls[row] = autos[row].isna().sum()
nulls   
    #add a key-value pair to the empty dictionary
# my_dictionary['name'] = "John Doe"

{'dateCrawled': 0,
 'name': 0,
 'seller': 0,
 'offerType': 0,
 'price': 0,
 'abtest': 0,
 'vehicleType': 5095,
 'yearOfRegistration': 0,
 'gearbox': 2680,
 'powerPS': 0,
 'model': 2758,
 'odometer': 0,
 'monthOfRegistration': 0,
 'fuelType': 4482,
 'brand': 0,
 'notRepairedDamage': 9829,
 'dateCreated': 0,
 'nrOfPictures': 0,
 'postalCode': 0,
 'lastSeen': 0}

In [78]:
percent_missing = autos.isnull().sum() * 100 / len(autos)
missing_value_autos = pd.DataFrame({'column_name': autos.columns,
                                 'percent_missing': percent_missing})
missing_value_autos

column_name  percent_missing
dateCrawled                  dateCrawled            0.000
name                                name            0.000
seller                            seller            0.000
offerType                      offerType            0.000
price                              price            0.000
abtest                            abtest            0.000
vehicleType                  vehicleType           10.190
yearOfRegistration    yearOfRegistration            0.000
gearbox                          gearbox            5.360
powerPS                          powerPS            0.000
model                              model            5.516
odometer                        odometer            0.000
monthOfRegistration  monthOfRegistration            0.000
fuelType                        fuelType            8.964
brand                              brand            0.000
notRepairedDamage      notRepairedDamage           19.658
dateCreated                  dateCreated            0.000
nrOfPictures                nrOfPictures            0.000
postalCode                    postalCode            0.000
lastSeen                        lastSeen            0.000

In [79]:
# looks at descriptive statistics for all columns, omiting missing values
autos.nunique()

dateCrawled            48213
name                   38754
seller                     2
offerType                  2
price                   2357
abtest                     2
vehicleType                8
yearOfRegistration        97
gearbox                    2
powerPS                  448
model                    245
odometer                  13
monthOfRegistration       13
fuelType                   7
brand                     40
notRepairedDamage          2
dateCreated               76
nrOfPictures               1
postalCode              7014
lastSeen               39481
dtype: int64

In [80]:
# finds unique values of columns, considering missing values
# for row in autos.columns:
#     print(row, autos[row].unique().shape)
    
# another way to check the same information: 
autos.nunique(dropna=False)

dateCrawled            48213
name                   38754
seller                     2
offerType                  2
price                   2357
abtest                     2
vehicleType                9
yearOfRegistration        97
gearbox                    3
powerPS                  448
model                    246
odometer                  13
monthOfRegistration       13
fuelType                   8
brand                     40
notRepairedDamage          3
dateCreated               76
nrOfPictures               1
postalCode              7014
lastSeen               39481
dtype: int64

In [81]:
# looks at descriptive statistics for all columns.
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-30 17:37:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  yearOfRegistration  gearbox       powerPS  model  \
count        44905        50000.000000    47320  50000.000000  47242   
unique           8                 NaN        2           NaN    245   
top      limousine                 NaN  manuell           NaN   golf   
freq         12859                 NaN    36993           NaN   4024   
mean           NaN         2005.073280      NaN    116.355920    NaN   
std            NaN          105.712813      NaN    209.216627    NaN   
min            NaN         1000.000000      NaN      0.000000    NaN   
25%            NaN         1999.000000      NaN     70.000000    NaN   
50%            NaN         2003.000000      NaN    105.000000    NaN   
75%            NaN         2008.000000      NaN    150.000000    NaN   
max            NaN         9999.000000      NaN  17700.000000    NaN   

         odometer  monthOfRegistration fuelType       brand notRepairedDamage  \
count       50000         50000.000000    45518       50000             40171   
unique         13                  NaN        7          40                 2   
top     150,000km                  NaN   benzin  volkswagen              nein   
freq        32424                  NaN    30107       10687             35232   
mean          NaN             5.723360      NaN         NaN               NaN   
std           NaN             3.711984      NaN         NaN               NaN   
min           NaN             0.000000      NaN         NaN               NaN   
25%           NaN             3.000000      NaN         NaN               NaN   
50%           NaN             6.000000      NaN         NaN               NaN   
75%           NaN             9.000000      NaN         NaN               NaN   
max           NaN            12.000000      NaN         NaN               NaN   

                dateCreated  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [82]:
# hist = autos.hist(figsize=(10,10),layout=(3,3))

## Notes about the dataset:
- The name of variables are in _camelCase_ - **TODO1**: transform them in _snake_case_.
- 'price' and 'odometer' are numerical variables stored as object: **TODO2**: transform them in _int_.
- There are values in German languages. Since it doesn't influence the results, I'm leaving them as they are.
- 'dateCrawled', 'dateCreated' and 'lastSeen'are stored as object - **TODO3**: convert to datetime
- 'postalCode'is stored as integer, which can cause errors ([look some explanations aboit it](https://streamofcoding.com/zip-codes-are-not-integers/)). The same for 'yearOfRegistration', 'power_P_S'and 'monthOfRegistration' - **TODO4**: convert them to _object_.
- 'nrOfPictures'has only one value, all records have 1 picture. - **TODO5**: remove this variable, explore deeper the 'seller', 'offerType', 'gearbox', 'notRepairedDamage' variables that seems to be booleans and keep with 'abtest' because it is expected to be a boolean.
- The variable with missing values are 'vehicleType'(10.19%), 'gearbox'(5.36%), model (5.516%), 'fuelType'(8.964%) and 'notRepairedDamage'(19.658%). Missing data can lead to a lack of precision in the statistical analysis. - **TODO6**: Once the missing values don't represent more than 20% of the values, replace them with the Modes.

**Other observations:**

> The price is the most important variable. Is there any variable strongly related to the price of the autos?


## Cleaning

### TODO1: writing style to use standard convention - convert columns from camelCase to snake_case.

snake_case is the writing style commonly used in Python creating a naming convention that avoids syntax errors.

In [83]:
# transforms the first letter in lowercase
# for the other letters, if there is some in uppercase, appends _ and lower
# otherwise, just keep with the letter in uppercase
#     for c in str[1:]: 

def change_case(str):
    res = [str[0].lower()] 
    for c in str[1:]: 
        if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'):
            res.append('_')
            res.append(c.lower())
        else:
            res.append(c)
    
    # uses the join methos to concatenate the letters
    return ''.join(res)
        

In [84]:
# defines each column value as a string and run the change_case function
new_columns = []
for c in autos.columns:
    str = c
    new_columns.append(change_case(str))
#     print(change_case(str))
# print(new_columns)

In [85]:
# defines the autos' columns using the list
autos.columns = new_columns
# list(autos.columns)
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type   price   abtest vehicle_type  year_of_registration    gearbox  \
0    Angebot  $5,000  control          bus                  2004    manuell   
1    Angebot  $8,500  control    limousine                  1997  automatik   

   power_p_s   model   odometer  month_of_registration fuel_type    brand  \
0        158  andere  150,000km                      3       lpg  peugeot   
1        286     7er  150,000km                      6    benzin      bmw   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                nein  2016-03-26 00:00:00               0        79588   
1                nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

### TODO2: about datatype - convert numerical variables from text to numbers

To prepare data to make math operations and sortings. 

In [86]:
# removes especial characters and renames columns
autos['price'] = autos['price'].str.replace('$', '').str.replace(',', '')
autos.rename({'price':'price_usd'}, inplace=True, axis=1)

autos['odometer'] = autos['odometer'].str.replace('km', '').str.replace(',', '')
autos.rename({'odometer':'odometer_km'}, inplace=True, axis=1)

autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type price_usd   abtest vehicle_type  year_of_registration    gearbox  \
0    Angebot      5000  control          bus                  2004    manuell   
1    Angebot      8500  control    limousine                  1997  automatik   

   power_p_s   model odometer_km  month_of_registration fuel_type    brand  \
0        158  andere      150000                      3       lpg  peugeot   
1        286     7er      150000                      6    benzin      bmw   

  not_repaired_damage         date_created  nr_of_pictures  postal_code  \
0                nein  2016-03-26 00:00:00               0        79588   
1                nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

In [87]:
# transforming 'odometer_km' and 'price_usd' in int
autos[['odometer_km', 'price_usd']] = autos[['odometer_km', 'price_usd']].astype(int)
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   date_crawled           50000 non-null  object
 1   name                   50000 non-null  object
 2   seller                 50000 non-null  object
 3   offer_type             50000 non-null  object
 4   price_usd              50000 non-null  int64 
 5   abtest                 50000 non-null  object
 6   vehicle_type           44905 non-null  object
 7   year_of_registration   50000 non-null  int64 
 8   gearbox                47320 non-null  object
 9   power_p_s              50000 non-null  int64 
 10  model                  47242 non-null  object
 11  odometer_km            50000 non-null  int64 
 12  month_of_registration  50000 non-null  int64 
 13  fuel_type              45518 non-null  object
 14  brand                  50000 non-null  object
 15  not_repaired_damage

### TODO3: about datatype - convert data about date and time to a proper format
The use of a proper format for date and time is needed to evaluate seasonalities and time-related patterns.

In [88]:
columns_to_datetime = ['date_crawled', 'date_created', 'last_seen']
for col in columns_to_datetime:
    autos[col] = pd.to_datetime(autos[col], format='%Y-%m-%d %H:%M:%S.%f')
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   date_crawled           50000 non-null  datetime64[ns]
 1   name                   50000 non-null  object        
 2   seller                 50000 non-null  object        
 3   offer_type             50000 non-null  object        
 4   price_usd              50000 non-null  int64         
 5   abtest                 50000 non-null  object        
 6   vehicle_type           44905 non-null  object        
 7   year_of_registration   50000 non-null  int64         
 8   gearbox                47320 non-null  object        
 9   power_p_s              50000 non-null  int64         
 10  model                  47242 non-null  object        
 11  odometer_km            50000 non-null  int64         
 12  month_of_registration  50000 non-null  int64         
 13  f

### TODO4: about datatype - convert ZIP code to text

ZIP code should be stored as text to avoid errors. To read more about it, click [here](https://streamofcoding.com/zip-codes-are-not-integers/).
Month, year and power ar also categoricals.

In [89]:
# transforming 'postal_code' in str
categorial_variables = ['postal_code', 'year_of_registration','power_p_s','month_of_registration']
for var in categorial_variables:
    autos[var] = pd.Series(autos[var], dtype='str')

autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   date_crawled           50000 non-null  datetime64[ns]
 1   name                   50000 non-null  object        
 2   seller                 50000 non-null  object        
 3   offer_type             50000 non-null  object        
 4   price_usd              50000 non-null  int64         
 5   abtest                 50000 non-null  object        
 6   vehicle_type           44905 non-null  object        
 7   year_of_registration   50000 non-null  object        
 8   gearbox                47320 non-null  object        
 9   power_p_s              50000 non-null  object        
 10  model                  47242 non-null  object        
 11  odometer_km            50000 non-null  int64         
 12  month_of_registration  50000 non-null  object        
 13  f

### TODO5: removing variables that do not add relevant information

- remove the variable for the number of pictures once all of the records have one picture
- explore deeper the 'seller', 'offer_type', 'gearbox', 'not_repaired_damage' variable that seem to be booleans 
- keep with 'abtest' because it is expected to be a boolean (test and control) 

After deeper analysis, I dropped columns with heavily one-sided valiables.

In [90]:
# drop columns with only one value (nrOfPictures)
for col in autos.columns:
    if len(autos[col].unique())==1:
        autos.drop(col, inplace = True, axis=1)
#         print(col)

In [91]:
n_categories = ['seller', 'offer_type', 'gearbox', 'not_repaired_damage']
for c in n_categories:
#     print(col, len(autos[col].value_counts()))
#     print(autos[column].unique().tolist())
    print(autos.nunique(dropna=False))

date_crawled             48213
name                     38754
seller                       2
offer_type                   2
price_usd                 2357
abtest                       2
vehicle_type                 9
year_of_registration        97
gearbox                      3
power_p_s                  448
model                      246
odometer_km                 13
month_of_registration       13
fuel_type                    8
brand                       40
not_repaired_damage          3
date_created                76
postal_code               7014
last_seen                39481
dtype: int64
date_crawled             48213
name                     38754
seller                       2
offer_type                   2
price_usd                 2357
abtest                       2
vehicle_type                 9
year_of_registration        97
gearbox                      3
power_p_s                  448
model                      246
odometer_km                 13
month_of_registration     

Finding the number of categories

The translation of these booleans is

['private', 'commercial']   
['Offer', 'Request']  
['manual', 'automatic', nan]  
['no', nan, 'yes']

Finding distinct values of **seller** and **offer_type**

In [92]:
percent_privat = len(autos.loc[autos['seller']== 'privat'])* 100 / len(autos)
print('The percentage of private seller is', percent_privat, '.')

The percentage of private seller is 99.998 .


In [93]:
percent_angebot = len(autos.loc[autos['offer_type']== 'Angebot'])* 100 / len(autos)
print('The percentage of private offer_type is', percent_angebot, '.')

The percentage of private offer_type is 99.998 .


In [94]:
# checking if the the seller commertial is the same with a different offer_type
autos.loc[autos['seller'] == 'gewerblich']

date_crawled                                         name  \
7738 2016-03-15 18:06:22  Verkaufe_mehrere_Fahrzeuge_zum_Verschrotten   

          seller offer_type  price_usd   abtest vehicle_type  \
7738  gewerblich    Angebot        100  control        kombi   

     year_of_registration  gearbox power_p_s   model  odometer_km  \
7738                 2000  manuell         0  megane       150000   

     month_of_registration fuel_type    brand not_repaired_damage  \
7738                     8    benzin  renault                 NaN   

     date_created postal_code           last_seen  
7738   2016-03-15       65232 2016-04-06 17:15:37

In [95]:
# checking if the the seller commertial is the same with a different offer_type
autos.loc[autos['offer_type'] == 'Gesuch']

date_crawled                  name  seller offer_type  price_usd  \
17541 2016-04-03 15:48:33  Suche_VW_T5_Multivan  privat     Gesuch          0   

      abtest vehicle_type year_of_registration gearbox power_p_s        model  \
17541   test          bus                 2005     NaN         0  transporter   

       odometer_km month_of_registration fuel_type       brand  \
17541       150000                     0       NaN  volkswagen   

      not_repaired_damage date_created postal_code           last_seen  
17541                 NaN   2016-04-03       29690 2016-04-05 15:16:06

The dataset has a majority of private sellers offering cars.
The only record with a non-private (commercial) seller is not the same one that has a different offer_type (request). It is not a record with mistakes. So, these variables are heavily one-sided and don't add helpful information to this analysis. So, they are dropped.

In [96]:
# drop columns with heavily one-sided valiables
print(autos.shape)
autos = autos.drop(['seller','offer_type'], axis=1)
print(autos.shape)

(50000, 19)
(50000, 17)


### TODO6: dealing with missing values - replace them with the Modes

Once the missing values don't represent more than 20% of the values, it is not needed to disregard these variables.
I filled them with the most occurred value instead. As the variables are categorical, filling in the missing values by Mode is the most indicated. To read more about it click [here](https://www.analyticsvidhya.com/blog/2021/10/handling-missing-value/).

In [97]:
# iterate over the columns of variables with missing values to check the most common values.
replace_to_mode = ['vehicle_type', 'gearbox', 'model', 'fuel_type', 'not_repaired_damage']
for var in replace_to_mode:
    print(var) 
    print(autos[var].value_counts().sort_values(ascending=False))   

vehicle_type
limousine     12859
kleinwagen    10822
kombi          9127
bus            4093
cabrio         3061
coupe          2537
suv            1986
andere          420
Name: vehicle_type, dtype: int64
gearbox
manuell      36993
automatik    10327
Name: gearbox, dtype: int64
model
golf          4024
andere        3528
3er           2761
polo          1757
corsa         1735
              ... 
kappa            2
200              1
b_max            1
i3               1
rangerover       1
Name: model, Length: 245, dtype: int64
fuel_type
benzin     30107
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64
not_repaired_damage
nein    35232
ja       4939
Name: not_repaired_damage, dtype: int64


Since all the variables with nulls values are categorical, they are replaced for the most frequently occurring value (mode). To read about it click [here](https://www.analyticsvidhya.com/blog/2021/10/handling-missing-value/).

In [98]:
#Replace the missing values for categorical columns with mode
replace_to_mode = ['vehicle_type', 'gearbox', 'model', 'fuel_type', 'not_repaired_damage']
# modes = {'vehicle_type':'limousine', 'gearbox': 'manuell', 'model_values': 'Golf', 'fuel_type':'benzin'}
 
for col in replace_to_mode:
    autos[col] = autos[col].fillna(autos[col].mode()[0])

autos.isnull().sum()

date_crawled             0
name                     0
price_usd                0
abtest                   0
vehicle_type             0
year_of_registration     0
gearbox                  0
power_p_s                0
model                    0
odometer_km              0
month_of_registration    0
fuel_type                0
brand                    0
not_repaired_damage      0
date_created             0
postal_code              0
last_seen                0
dtype: int64

In [99]:
fuel_type_values = autos['fuel_type'].value_counts()
fuel_type_values.sort_values(ascending=False)

benzin     34589
diesel     14567
lpg          691
cng           75
hybrid        37
andere        22
elektro       19
Name: fuel_type, dtype: int64

In [100]:
not_repaired_damage_values = autos['not_repaired_damage'].value_counts()
not_repaired_damage_values.sort_values(ascending=False)

nein    45061
ja       4939
Name: not_repaired_damage, dtype: int64

In [101]:
# finds unique values of columns
for row in autos.columns:
    print(row, autos[row].unique().shape)

date_crawled (48213,)
name (38754,)
price_usd (2357,)
abtest (2,)
vehicle_type (8,)
year_of_registration (97,)
gearbox (2,)
power_p_s (448,)
model (245,)
odometer_km (13,)
month_of_registration (13,)
fuel_type (7,)
brand (40,)
not_repaired_damage (2,)
date_created (76,)
postal_code (7014,)
last_seen (39481,)


In [102]:
# autos.info()

In [103]:
# transforming 'odometer_km' and 'price_usd' in int
autos[['odometer_km', 'price_usd']] = autos[['odometer_km', 'price_usd']].astype(int)
# autos.info()

In [104]:
autos.head()

date_crawled                                               name  \
0 2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1 2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2 2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3 2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4 2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price_usd   abtest vehicle_type year_of_registration    gearbox power_p_s  \
0       5000  control          bus                 2004    manuell       158   
1       8500  control    limousine                 1997  automatik       286   
2       8990     test    limousine                 2009    manuell       102   
3       4350  control   kleinwagen                 2007  automatik        71   
4       1350     test        kombi                 2003    manuell         0   

    model  odometer_km month_of_registration fuel_type       brand  \
0  andere       150000                     3       lpg     peugeot   
1     7er       150000                     6    benzin         bmw   
2    golf        70000                     7    benzin  volkswagen   
3  fortwo        70000                     6    benzin       smart   
4   focus       150000                     7    benzin        ford   

  not_repaired_damage date_created postal_code           last_seen  
0                nein   2016-03-26       79588 2016-04-06 06:45:54  
1                nein   2016-04-04       71034 2016-04-06 14:45:08  
2                nein   2016-03-26       35394 2016-04-06 20:15:37  
3                nein   2016-03-12       33729 2016-03-15 03:16:28  
4                nein   2016-04-01       39218 2016-04-01 14:38:50

## Extra cleaning

- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns



In [105]:
german_cols = ['vehicle_type', 'gearbox', 'fuel_type', 'not_repaired_damage']

for x in german_cols:
    print(x) 
    print(autos[x].unique())

# # for col in german_cols:
# # #     print(col, len(autos[col].value_counts()))
# #     print(autos[col].unique().tolist())
print('lpg = Liquefied petroleum gas') 
print('cng = Compressed natural gas')  
print('suv = Sports Utility Vehicle')
print('cabrio = convertible')

vehicle_type
['bus' 'limousine' 'kleinwagen' 'kombi' 'coupe' 'suv' 'cabrio' 'andere']
gearbox
['manuell' 'automatik']
fuel_type
['lpg' 'benzin' 'diesel' 'cng' 'hybrid' 'elektro' 'andere']
not_repaired_damage
['nein' 'ja']
lpg = Liquefied petroleum gas
cng = Compressed natural gas
suv = Sports Utility Vehicle
cabrio = convertible


In [106]:
# a function to translate words
def translator_column(df,column,value_dict):
    '''translates values in df column using value_dict as reference'''
    df[column] = df[column].map(value_dict)


# A dictionary of translations
translations ={# vehicle type terms
               'bus':'bus',
               'limousine':'limousine',
               'kleinwagen': 'small car',
               'kombi': 'kombi',
               'coupe':'coupe',
               'suv':'suv',
               'cabrio': 'cabrio',
               'andere': 'other',
               'cabrio': 'convertible',
               # gear box terms
               'manuell': 'manual',
               'automatik': 'automatic',
               # fuel type terms
               'lpg':'lpg',
               'benzin': 'gasoline',
               'diesel':'diesel',
               'cng':'cng',
               'hybrid':'hybrid',
               'elektro':'electric',
               'andere':'other',
               # unrepaired damage terms
               'nein':'no',
               'ja':'yes'       
}    
    
for i in german_cols:     
    translator_column(autos,i,translations)    
 

In [107]:
german_cols = ['vehicle_type', 'gearbox', 'fuel_type', 'not_repaired_damage']

for x in german_cols:
    print(x) 
    print(autos[x].unique())

vehicle_type
['bus' 'limousine' 'small car' 'kombi' 'coupe' 'suv' 'convertible' 'other']
gearbox
['manual' 'automatic']
fuel_type
['lpg' 'gasoline' 'diesel' 'cng' 'hybrid' 'electric' 'other']
not_repaired_damage
['no' 'yes']


## Removing outliers

Outliers are occurrences with values that are very different from the others, which are usually unrealistic, and which generate false trends in statistical analyses.

The outliers of the numerical variables 'odometer_km' and also of 'price_usd' and 'year_of_registration' are found using interquatile ranges (IQR) and droped. Using quartiles, the sample is divided into 4 equal parts (quartiles Q1,Q2,Q3 and Q4), we use the central parts, considering a gap of 1.5% for more in the upper limit and for less in the lower limit. Occurrences outside the bounds plus or less gaps are removed from the dataframe.

>Q1: 25% of the occurences  
Q2: 50% of the occurences  
Q3: 75% of the occurences  
IQR = Q3 − Q1  
inferior limit = (Q1-1,5%) * IQR   
superior limit = (Q3+1,5%) * IQR

The outliers for 'year_of_registration' were detected by empirical knolewdge. Dates before 1900 and after 2016 are outliers because cars where invented in 1900 and the dates for registration were in 2016, so the cars can't be olders or newers than those years.

In [108]:
odometer_price = ['odometer_km', 'price_usd']
for element in odometer_price:
    print(element)
    print(autos[element].value_counts().sort_index(ascending=False).head(15))

odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64
price_usd
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price_usd, dtype: int64


In [109]:
autos[['odometer_km','price_usd']].astype(int).describe()

odometer_km     price_usd
count   50000.000000  5.000000e+04
mean   125732.700000  9.840044e+03
std     40042.211706  4.811044e+05
min      5000.000000  0.000000e+00
25%    125000.000000  1.100000e+03
50%    150000.000000  2.950000e+03
75%    150000.000000  7.200000e+03
max    150000.000000  1.000000e+08

In [110]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [111]:
# categorical variables
cat_autos = autos.select_dtypes(include = np.object)
cat_autos.head()

name   abtest vehicle_type  \
0                   Peugeot_807_160_NAVTECH_ON_BOARD  control          bus   
1         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  control    limousine   
2                         Volkswagen_Golf_1.6_United     test    limousine   
3  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  control    small car   
4  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...     test        kombi   

  year_of_registration    gearbox power_p_s   model month_of_registration  \
0                 2004     manual       158  andere                     3   
1                 1997  automatic       286     7er                     6   
2                 2009     manual       102    golf                     7   
3                 2007  automatic        71  fortwo                     6   
4                 2003     manual         0   focus                     7   

  fuel_type       brand not_repaired_damage postal_code  
0       lpg     peugeot                  no       79588  
1  gasoline         bmw                  no       71034  
2  gasoline  volkswagen                  no       35394  
3  gasoline       smart                  no       33729  
4  gasoline        ford                  no       39218

In [112]:
numerics = ['int64', 'float64']
autos_num = autos.select_dtypes(include=numerics)
autos_num


price_usd  odometer_km
0           5000       150000
1           8500       150000
2           8990        70000
3           4350        70000
4           1350       150000
...          ...          ...
49995      24900       100000
49996       1980       150000
49997      13200         5000
49998      22900        40000
49999       1250       150000

[50000 rows x 2 columns]

In [113]:
print('Frequency of values for odometer_km')
autos['odometer_km'].value_counts()

Frequency of values for odometer_km


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

The frequency of 150,000km (32,424) is really higher than of 125,000km (5,170), probably many of these autos have much more than 150,000km and records are inacurrate. 
It also makes its distribution heavily one-sided.

In [114]:
autos.shape

(50000, 17)

In [115]:
autos['price_usd'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [116]:
# removes outliers from prices_usd

q1 = autos['price_usd'].quantile(0.25)
q3 = autos['price_usd'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - 1.5 * iqr
upper_bound = q3 + 1.5 * iqr
autos = autos[(autos['price_usd'] >= lower_bound) & (autos['price_usd'] <= upper_bound)]
                                                                 
print(q1)
print(q3)
print(iqr)

print(lower_bound)
print(upper_bound)


1100.0
7200.0
6100.0
-8050.0
16350.0


In [117]:
# autos.shape
# autos['price_usd'].describe()

autos['price_usd'].value_counts(normalize=True).head()

0       0.030747
500     0.016899
1500    0.015882
2500    0.013913
1000    0.013826
Name: price_usd, dtype: float64

In [118]:
# plt.rcParams.update({'figure.figsize':(5,3), 'figure.dpi':100})
# plt.xlim([0,17000])
# plt.hist(autos['price_usd'], bins=50)
# plt.gca().set(title='Frequency Histogram', ylabel='Frequency',  xlabel='Price_usd')
# plt.show()

To remove these outliers we have to return the serie to integer.

In [119]:
autos['year_of_registration'].describe()
autos['year_of_registration'].value_counts(normalize=True).head()

2000    0.072269
1999    0.064675
2005    0.064199
2003    0.058356
2004    0.058292
Name: year_of_registration, dtype: float64

In [120]:
autos['year_of_registration'] = pd.Series(autos['year_of_registration'], dtype='int')

In [121]:
#removes the outliers in year_of_registration and check the distribution
earlier_year = 1900
later_year = 2016
autos = autos[(autos['year_of_registration'] >= earlier_year) & (autos['year_of_registration'] <= later_year)]
# autos['year_of_registration'].max()
# autos['year_of_registration'].min()
dist_years = autos['year_of_registration'].value_counts(normalize=True)
dist_years 

2000    0.075405
1999    0.067481
2005    0.066984
2003    0.060889
2004    0.060821
          ...   
1929    0.000023
1941    0.000023
1953    0.000023
1938    0.000023
1952    0.000023
Name: year_of_registration, Length: 70, dtype: float64

In [122]:
# data=autos['year_of_registration'] 
# plt.rcParams.update({'figure.figsize':(8,5), 'figure.dpi':100})
# plt.locator_params(axis = 'x', nbins = 15)
# plt.hist(data, bins=50)
# plt.gca().set(title='Frequency Histogram', ylabel='Frequency')
# plt.show()

## date and time

In [123]:
autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled date_created           last_seen
0 2016-03-26 17:47:46   2016-03-26 2016-04-06 06:45:54
1 2016-04-04 13:38:56   2016-04-04 2016-04-06 14:45:08
2 2016-03-26 18:57:24   2016-03-26 2016-04-06 20:15:37
3 2016-03-12 16:58:10   2016-03-12 2016-03-15 03:16:28
4 2016-04-01 14:38:50   2016-04-01 2016-04-01 14:38:50

In [124]:
# columns_to_datetime = ['date_crawled', 'date_created', 'last_seen']
# for col in columns_to_datetime:
#     autos[col] = pd.to_datetime(autos[col], format='%Y-%m-%d %H:%M:%S.%f')
# autos.info()

In [125]:
autos['just_date_date_crawled'] = autos['date_crawled'].dt.date
autos['just_date_date_created'] = autos['date_created'].dt.date
autos['just_date_last_seen'] = autos['last_seen'].dt.date
# autos['just_date_date_crawled']

In [126]:
autos.head(2)

date_crawled                                        name  price_usd  \
0 2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD       5000   
1 2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik       8500   

    abtest vehicle_type  year_of_registration    gearbox power_p_s   model  \
0  control          bus                  2004     manual       158  andere   
1  control    limousine                  1997  automatic       286     7er   

   odometer_km month_of_registration fuel_type    brand not_repaired_damage  \
0       150000                     3       lpg  peugeot                  no   
1       150000                     6  gasoline      bmw                  no   

  date_created postal_code           last_seen just_date_date_crawled  \
0   2016-03-26       79588 2016-04-06 06:45:54             2016-03-26   
1   2016-04-04       71034 2016-04-06 14:45:08             2016-04-04   

  just_date_date_created just_date_last_seen  
0             2016-03-26          2016-04-06  
1             2016-04-04          2016-04-06

In [127]:
datetimes_autos = ['just_date_date_crawled','just_date_date_created','just_date_last_seen']

for var in datetimes_autos:
    print(var, ': n of unique dates =', autos[var].unique().size)
    print(autos[var]
          .value_counts(normalize=True, dropna=False)
          .round(3)
          .sort_index(ascending=False))

just_date_date_crawled : n of unique dates = 34
2016-04-07    0.001
2016-04-06    0.003
2016-04-05    0.013
2016-04-04    0.037
2016-04-03    0.039
2016-04-02    0.035
2016-04-01    0.033
2016-03-31    0.032
2016-03-30    0.034
2016-03-29    0.034
2016-03-28    0.035
2016-03-27    0.030
2016-03-26    0.033
2016-03-25    0.032
2016-03-24    0.029
2016-03-23    0.033
2016-03-22    0.032
2016-03-21    0.037
2016-03-20    0.038
2016-03-19    0.034
2016-03-18    0.013
2016-03-17    0.032
2016-03-16    0.030
2016-03-15    0.034
2016-03-14    0.037
2016-03-13    0.016
2016-03-12    0.037
2016-03-11    0.032
2016-03-10    0.033
2016-03-09    0.033
2016-03-08    0.034
2016-03-07    0.036
2016-03-06    0.014
2016-03-05    0.025
Name: just_date_date_crawled, dtype: float64
just_date_date_created : n of unique dates = 71
2016-04-07    0.001
2016-04-06    0.003
2016-04-05    0.012
2016-04-04    0.037
2016-04-03    0.039
              ...  
2015-12-30    0.000
2015-12-05    0.000
2015-11-10    0.000

In [128]:
pd.set_option('display.max_rows', 100) # increase the maximum number of rows to visualise all rows and later reset it

In [129]:
# 'date_created'
print('date_created', ': n of unique dates =', autos['just_date_date_created'].unique().size)
print(autos['just_date_date_created']
          .value_counts(normalize=True, dropna=False)
          .round(3)
          .sort_index(ascending=False))

date_created : n of unique dates = 71
2016-04-07    0.001
2016-04-06    0.003
2016-04-05    0.012
2016-04-04    0.037
2016-04-03    0.039
2016-04-02    0.035
2016-04-01    0.033
2016-03-31    0.032
2016-03-30    0.034
2016-03-29    0.034
2016-03-28    0.035
2016-03-27    0.030
2016-03-26    0.033
2016-03-25    0.032
2016-03-24    0.029
2016-03-23    0.033
2016-03-22    0.032
2016-03-21    0.037
2016-03-20    0.038
2016-03-19    0.033
2016-03-18    0.014
2016-03-17    0.032
2016-03-16    0.030
2016-03-15    0.034
2016-03-14    0.035
2016-03-13    0.017
2016-03-12    0.037
2016-03-11    0.033
2016-03-10    0.032
2016-03-09    0.033
2016-03-08    0.034
2016-03-07    0.035
2016-03-06    0.015
2016-03-05    0.023
2016-03-04    0.001
2016-03-03    0.001
2016-03-02    0.000
2016-03-01    0.000
2016-02-29    0.000
2016-02-28    0.000
2016-02-27    0.000
2016-02-26    0.000
2016-02-25    0.000
2016-02-24    0.000
2016-02-23    0.000
2016-02-22    0.000
2016-02-21    0.000
2016-02-20    0.000
20

In [130]:
pd.reset_option('display.max_rows')   # reset the maximum number of rows

### Partial findings:
#### date crewled: When this ad was first crawled. All field-values are taken from this date.
- The ads crewled were collected on a daily basis from 05-03-2016 to 07-04-2016, having records for 34 days.
- The overall distribuiton of ads collected kept between 2% and 3%.
- Drops were observed on March (6th, 13th, 18th) that were weekend or close (Friday).

#### date created: The date on which the eBay listing was created.

- The ads were created from 03-03-2016 and occurences were registered until 07-04-2016, having non-zero records for 34 days.
- The overall distribuiton of ads collected kept between 2% and 3% (mainly 3%).
- A drop in occurrences was observed March (6th, 13th, 18th).

#### last seen: When the crawler saw this ad last online.
- The ads crewled were collected on a daily basis from 05-03-2016 to 07-04-2016, having records for 34 days.
- The distribution of ads seen has been gradually increasing
- Slight peaks were observed in March (12nd, 17th), as well as a sharp peak in April (5th, 6th). Shortly after the peaks, gradual declines and resumptions are observed.

## Exploring the brands

In [131]:
# select the top6 brands
# autos['brand'].unique()
# pct_brands = autos['brand'].value_counts(normalize=True, ascending=False, dropna=False)
top_brands = (autos['brand']
             .value_counts(normalize=True, ascending=False, dropna=False)
             .head(6)
             .index)
top_brands

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')

The brands with six higher occurrence are volkswagen(21%), opel(10%), bmw(10%), mercedes_benz(9%), audi(8%) and ford(6%). I chose the brands with more than 5% of the occurrences.   

### Aggregate price and mileage based on brands

In [132]:
brand_mean_prices = {}
for brand in top_brands:
    prices = autos['price_usd'][autos['brand'] == brand]
#     print(brand, prices)
    mean_price = prices.mean()
    brand_mean_prices[brand] = int(mean_price) 
#     print(brand)
#     print(mean_price)
        
print(brand_mean_prices)

price_series = pd.Series(brand_mean_prices)
price_series


{'volkswagen': 4041, 'opel': 2621, 'bmw': 5469, 'mercedes_benz': 5156, 'audi': 5524, 'ford': 2865}


volkswagen       4041
opel             2621
bmw              5469
mercedes_benz    5156
audi             5524
ford             2865
dtype: int64

In [133]:
price_df = pd.DataFrame(price_series, columns=['mean_price_usd'])
price_df

mean_price_usd
volkswagen               4041
opel                     2621
bmw                      5469
mercedes_benz            5156
audi                     5524
ford                     2865

Audi shown to have a higher price, followed by BMW, Mercedes Benz, Volkswagen, Ford and Opel.

In [134]:
aggregate_mileage = {}
for brand in top_brands:
    mileages = autos['odometer_km'][autos['brand'] == brand]
    mean_mileage_km = mileages.mean()
    aggregate_mileage[brand] = int(mean_mileage_km) 
#     print(brand)
#     print(mean_price)
        
print(aggregate_mileage)

mileage_series = pd.Series(aggregate_mileage)
mileage_series

{'volkswagen': 132692, 'opel': 130371, 'bmw': 138471, 'mercedes_benz': 138341, 'audi': 139527, 'ford': 126726}


volkswagen       132692
opel             130371
bmw              138471
mercedes_benz    138341
audi             139527
ford             126726
dtype: int64

In [135]:
price_df['mean_mileage_km'] = mileage_series
price_df

mean_price_usd  mean_mileage_km
volkswagen               4041           132692
opel                     2621           130371
bmw                      5469           138471
mercedes_benz            5156           138341
audi                     5524           139527
ford                     2865           126726